nx/lib/nx/binary_backend.ex

```diff
@@ -1679,6 +1679,32 @@ defmodule Nx.BinaryBackend do
     from_binary(out, data)
   end
 
+  @impl true
+  def take(out, tensor, indices, axis) do
+    %T{type: {_, idx_size}} = indices
+
+    # We iterate over the indices in a flat manner,
+    # and take a unit tensor slice along axis given
+    # by each index. Then we concatenate the tensors
+    # along the axis, which gives us the result with
+    # index dimensions flattened and we just reshape.
+
+    tensor_rank = tuple_size(tensor.shape)
+    slice_start = List.duplicate(0, tensor_rank)
+    slice_lengths = tensor.shape |> Tuple.to_list() |> List.replace_at(axis, 1)
+
+    slices =
+      for <<bin::size(idx_size)-bitstring <- to_binary(indices)>> do
+        idx = binary_to_number(bin, indices.type)
+        slice_start = List.replace_at(slice_start, axis, idx)
+        Nx.slice(tensor, slice_start, slice_lengths)
+      end
+
+    slices
+    |> Nx.concatenate(axis: axis)
+    |> Nx.reshape(out.shape, names: out.names)
```

Comment:
This does work, but is certainly less neat. I still wonder what's wrong with using `Nx` calls, we know the tensor backend will be `BinaryBackend` and the call will be dispatched here, so it should be effectively the same.

```suggestion
    slices =
      for <<bin::size(idx_size)-bitstring <- to_binary(indices)>> do
        idx = binary_to_number(bin, indices.type)
        slice_start = List.replace_at(slice_start, axis, idx)
        strides = List.duplicate(1, tensor_rank)
        slice_shape = List.to_tuple(slice_lengths)
        slice(%{tensor | shape: slice_shape}, tensor, slice_start, slice_lengths, strides)
      end

    concat_shape = put_elem(tensor.shape, axis, length(slices))
    result = concatenate(%{tensor | shape: concat_shape}, slices, axis)
    reshape(out, result, out.shape)
```

Reply:
If we're to refactor, I'd drop down to a lower level, manipulating the binary directly, instead of building tensors.
However, neither me nor @josevalim can seem to remember the reasoning to _not_ call Nx.
Since there isn't much in the way of efficiency loss here, I particularly see no problem in keeping your code as is.

Reply:
FWIW there should be no overhead with packing/unpacking tensors and binaries, since they are kept as such.

Reply:
Yeah. I'll mark these comments as resolved :)

Reply:
@jonatanklosko yeah, the goal to drop to a lower level is that we wouldn’t compute the out tensor, we would simply call whatever slice call internally. If there is nothing internal, we should extract it out.

I think the answer to our question will depend on the grad for take. We can see the grad in Jax. If the grad is build on top of concatenate, then this should be left as is and we would try to move it upstream. However, if this is meant to have its own operation, with its own grad, I think making it independent of Nx is a better design.

Reply:
The gradient uses scatter_add, so I would like to see a version of this that doesn't call Nx for us to compare it with. :)

Reply:
Alright, will do! Also, are the jax gradient rules in [lax.py](https://github.com/google/jax/blob/main/jax/_src/lax/lax.py) or somewhere else? The only relevant thing I found was [gather gradient](https://github.com/google/jax/blob/main/jax/_src/lax/lax.py#L4581-L4585), which would suggest gradient is computed on a lower level, but this snippet doesn't look like an actual implementation.

Reply:
@jonatanklosko it is not always straightforward to find but my best guess is that take is implemented on top of gather, yeah.

```
>>> jax.make_jaxpr(jax.grad(lambda x: jnp.sum(jnp.take(x, [0, 1, 2]))))(jnp.array([[1.0, 2, 3], [4, 5, 6]]))
{ lambda a ; b.
  let c = reshape[ dimensions=None
                   new_sizes=(6,) ] b
      d = broadcast_in_dim[ broadcast_dimensions=(0,)
                            shape=(3, 1) ] a
      e = gather[ dimension_numbers=GatherDimensionNumbers(offset_dims=(), collapsed_slice_dims=(0,), start_index_map=(0,))
                  slice_sizes=(1,) ] c d
      _ = reduce_sum[ axes=(0,) ] e
      f = broadcast_in_dim[ broadcast_dimensions=(  )
                            shape=(3,) ] 1.0
      g = broadcast_in_dim[ broadcast_dimensions=(  )
                            shape=(  ) ] 0.0
      h = broadcast_in_dim[ broadcast_dimensions=(  )
                            shape=(6,) ] g
      i = scatter-add[ dimension_numbers=ScatterDimensionNumbers(update_window_dims=(), inserted_window_dims=(0,), scatter_dims_to_operand_dims=(0,))
                       indices_are_sorted=False
                       unique_indices=False
                       update_consts=(  )
                       update_jaxpr={ lambda  ; a b.
                                      let c = add a b
                                      in (c,) } ] h d f
      j = reshape[ dimensions=None
                   new_sizes=(2, 3) ] i
  in (j,) }
```

Reply:
The `take` implementation is [here](https://github.com/google/jax/blob/f7009ed1072bf48ea6ba9d6d2ce573ce0ce13377/jax/_src/numpy/lax_numpy.py#L4779-L4780) and it definitely uses gather, I wasn't sure if the grad is implemented for take separately or rather a generic one for `lax.gather`.

Reply:
@josevalim I refactored `slice` and `concatenate` into `do_slice` and `do_concatenate` respectively and rewrote take in terms of these.

Reply:
Looking good. It ended up not being much more verbose than the Nx-calls version

Reply:
Yup, that was a good call! :tea: